File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
from collections import defaultdict

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        
        dd = defaultdict(int)
        prev_ptr = 0
        
        size = len(s)
        ans = 0
        
        for idx in range(size):
            
            char = s[idx]
            dd[char] +=1
            
            while(prev_ptr <= idx and dd[char] > 1):
                
                prev_char = s[prev_ptr] 
                prev_ptr +=1
                
                if (dd[prev_char] == 1):
                    dd.pop(prev_char)
                
                else:
                    dd[prev_char] -=1
            
            curr_len = idx - prev_ptr+1
            
            # print(idx ,prev_ptr , curr_len)
            ans = max(ans , curr_len)
        
        return (ans)

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.rs
use std::collections::HashMap;
use std::cmp::max;

impl Solution {

    pub fn length_of_longest_substring(s: String) -> i32 {
    let mut counting:HashMap<char, u8> = HashMap::new();

    let mut ans = 0;

    let mut prev = 0;

    let ss:Vec<char> = s.chars().collect();
    let mut curr_len;


    for idx in 0..ss.len(){
        
        let  curr_char = ss[idx];

        if counting.contains_key(&curr_char){

            while prev < idx && counting.contains_key(&curr_char) {

                let count = counting.entry(ss[prev]).or_insert(0);
                
                counting.remove(&ss[prev]);

                prev+=1;

            }
        }
        counting.insert(curr_char ,1);

        curr_len = idx - prev+1;

        ans = max(ans , curr_len as i32)
    }    
    ans 
}

}

